### PR TITLE
Updated to Loop to load cgm plugin

### PR DESCRIPTION
**Pull request by nhamming**
https://tidepool.atlassian.net/browse/LOOP-317

Updates to load the Dexcom CGM kit plugin into Tidepool Loop.

### CODE REVIEW

File: Loop/Extensions/UIAlertController.swift [hunk 1]
```diff
@@ -72,35 +72,35 @@ extension UIAlertController {
...
-    ///   - cgmManagers: An array of CGMManager-conforming types
+    ///   - cgmManagers: An array of CGMManagers
     ///   - pumpManager: A PumpManager/CGMManager combo instance
     ///   - selectionHandler: A closure to execute when either a new CGMManager or the current PumpManager is selected
-    ///   - cgmManager: The selected CGMManager type
-    ///   - pumpManager: The selected PumpManager instance
-    internal convenience init(cgmManagers: [CGMManagerUI.Type], pumpManager: CGMManager?, selectionHandler: @escaping (_ cgmManager: CGMManagerUI.Type?, _ pumpManager: CGMManager?) -> Void) {
+    ///   - cgmManager: The selected CGMManager
+    ///   - pumpManager: The selected PumpManager
+    internal convenience init(cgmManagers: [AvailableDevice], pumpManager: CGMManager?, selectionHandler: @escaping (_ identifier: String, _ pumpManager: CGMManager?) -> Void) {
```
Comment (darinkrauss) [inline Loop/Extensions/UIAlertController.swift]: Why the switch to `AvailableDevice`?

File: Loop/Extensions/UIAlertController.swift [hunk 2]
```diff
@@ -72,35 +72,35 @@ extension UIAlertController {
...
-
+        
         if let pumpManager = pumpManager {
             addAction(UIAlertAction(
                 title: pumpManager.localizedTitle,
                 style: .default,
                 handler: { (_) in
-                    selectionHandler(nil, pumpManager)
-                }
+                    selectionHandler("", pumpManager)
```
Comment (darinkrauss) [inline Loop/Extensions/UIAlertController.swift]: Does the empty string mean 1) no selection, or 2) an empty string selection? (I've never been fond of using an empty string to indicate "no choice". It's along the lines of using zero to indicate "no choice", when in fact zero is a valid number.)

File: Loop/Managers/DeviceDataManager.swift [hunk 3]
```diff
@@ -147,6 +150,32 @@ final class DeviceDataManager {
...
+        guard let managerIdentifier = rawValue["managerIdentifier"] as? String else {
+            return nil
+        }
+        
+        return cgmManagerTypeByIdentifier(managerIdentifier)
+    }
+    
+    func cgmManagerFromRawValue(_ rawValue: [String: Any]) -> CGMManagerUI? {
+        guard let rawState = rawValue["state"] as? CGMManager.RawStateValue,
+            let Manager = cgmManagerTypeFromRawValue(rawValue)
```
Comment (darinkrauss) [inline Loop/Managers/DeviceDataManager.swift:169]: [nit] `Manager` is a variable of type `Type`. Should it be capitalized or not?

File: Loop/Plugins/LoopPlugins.swift [hunk 4]
```diff
@@ -67,6 +67,43 @@ class PluginManager {
...
+                            return plugin.cgmManagerType
+                        } else {
+                            fatalError("PrincipalClass does not conform to LoopUIPlugin")
+                        }
+                        
+                    } else {
+                        fatalError("PrincipalClass not found")
+                    }
+                } catch let error {
+                    print(error)
```
Comment (darinkrauss) [inline Loop/Plugins/LoopPlugins.swift:88]: [nit] For debugging or permanent?

File: Loop/Plugins/LoopPlugins.swift [hunk 5]
```diff
@@ -67,6 +67,43 @@ class PluginManager {
...
+    
+    var availableCGMManagers: [AvailableDevice] {
+        return pluginBundles.compactMap({ (bundle) -> AvailableDevice? in
+            guard let title = bundle.object(forInfoDictionaryKey: LoopPluginBundleKey.cgmManagerDisplayName.rawValue) as? String,
+                let identifier = bundle.object(forInfoDictionaryKey: LoopPluginBundleKey.cgmManagerIdentifier.rawValue) as? String else {
+                    return nil
+            }
+            
+            return AvailableDevice(identifier: identifier, localizedTitle: title)
+
```
Comment (darinkrauss) [inline Loop/Plugins/LoopPlugins.swift]: [nit] Extra line.

File: Loop/View Controllers/SettingsTableViewController.swift [hunk 6]
```diff
@@ -429,17 +429,17 @@ final class SettingsTableViewController: UITableViewController {
                 present(alert, animated: true, completion: nil)
             } else {
                 // Add new CGM
-                let cgmManagers = allCGMManagers.compactMap({ $0 as? CGMManagerUI.Type })
+                let cgmManagers = dataManager.availableCGMManagers
 
                 switch cgmManagers.count {
                 case 1:
-                    if let CGMManagerType = cgmManagers.first {
+                    if let cgmManager = cgmManagers.first, let CGMManagerType = dataManager.cgmManagerTypeByIdentifier(cgmManager.identifier) {
```
Comment (darinkrauss) [inline Loop/View Controllers/SettingsTableViewController.swift:436]: [nit] Capitalization of variable?

File: Loop/View Controllers/SettingsTableViewController.swift [hunk 7]
```diff
@@ -429,17 +429,17 @@ final class SettingsTableViewController: UITableViewController {
...
                         setupCGMManager(CGMManagerType, indexPath: indexPath)
                     }
                 case let x where x > 1:
-                    let alert = UIAlertController(cgmManagers: cgmManagers, pumpManager: dataManager.pumpManager as? CGMManager) { [weak self] (cgmManager, pumpManager) in
-                        if let CGMManagerType = cgmManager {
-                            self?.setupCGMManager(CGMManagerType, indexPath: indexPath)
+                    let alert = UIAlertController(cgmManagers: cgmManagers, pumpManager: dataManager.pumpManager as? CGMManager) { [weak self] (identifier, pumpManager) in
+                        if let self = self, let CGMManagerType = self.dataManager.cgmManagerTypeByIdentifier(identifier) {
```
Comment (darinkrauss) [inline Loop/View Controllers/SettingsTableViewController.swift]: Consider:

```
if let self = self {
    if let ... {
    } else if let ... {
    }
}

**Review by darinkrauss — APPROVED** [review]
A few questions and minor comments, but nothing show stopping. :shipit:

File: Loop/Extensions/UIAlertController.swift (hunk 1 above)
Comment (nhamming) [reply Loop/Extensions/UIAlertController.swift]: This allows for both static CGMManagers and those loaded by the newer plugin architecture.

File: Loop/Managers/DeviceDataManager.swift (hunk 3 above)
Comment (nhamming) [reply Loop/Managers/DeviceDataManager.swift:169]: I'm following the pattern for the pump managers. See line 164.

File: Loop/Plugins/LoopPlugins.swift (hunk 4 above)
Comment (nhamming) [reply Loop/Plugins/LoopPlugins.swift:88]: Again follow pump manager pattern (line 52). But I have been using this for debug.

File: Loop/Plugins/LoopPlugins.swift (hunk 5 above)
Comment (nhamming) [reply Loop/Plugins/LoopPlugins.swift]: will remove

File: Loop/View Controllers/SettingsTableViewController.swift (hunk 7 above)
Comment (nhamming) [reply Loop/View Controllers/SettingsTableViewController.swift]: I agree. will do

File: Loop/Extensions/UIAlertController.swift (hunk 2 above)
Comment (nhamming) [reply Loop/Extensions/UIAlertController.swift]: I agree. Good catch. Will update to `String?` to allow return of `nil`

File: Loop/View Controllers/SettingsTableViewController.swift (hunk 6 above)
Comment (nhamming) [reply Loop/View Controllers/SettingsTableViewController.swift:436]: In the end it is used for initialization. So it is a class like. Again, just trying to follow existing patterns.

**Review by darinkrauss — APPROVED** [review]
LGTM after latest commit. :rocket:

**Pull request merged**